onmt/translate/greedy_search.py

```diff
@@ -68,38 +68,60 @@ class RandomSampling(DecodeStrategy):
         bos (int): See base.
         eos (int): See base.
         batch_size (int): See base.
-        device (torch.device or str): See base ``device``.
         min_length (int): See base.
         max_length (int): See base.
         block_ngram_repeat (int): See base.
         exclusion_tokens (set[int]): See base.
         return_attention (bool): See base.
         max_length (int): See base.
         sampling_temp (float): See
-            :func:`~onmt.translate.random_sampling.sample_with_temperature()`.
+            :func:`~onmt.translate.greedy_search.sample_with_temperature()`.
         keep_topk (int): See
-            :func:`~onmt.translate.random_sampling.sample_with_temperature()`.
-        memory_length (LongTensor): Lengths of encodings. Used for
-            masking attention.
+            :func:`~onmt.translate.greedy_search.sample_with_temperature()`.
     """
 
-    def __init__(self, pad, bos, eos, batch_size, device,
-                 min_length, block_ngram_repeat, exclusion_tokens,
-                 return_attention, max_length, sampling_temp, keep_topk,
-                 memory_length):
-        super(RandomSampling, self).__init__(
-            pad, bos, eos, batch_size, device, 1,
-            min_length, block_ngram_repeat, exclusion_tokens,
-            return_attention, max_length)
+    def __init__(self, pad, bos, eos, batch_size, min_length,
+                 block_ngram_repeat, exclusion_tokens, return_attention,
+                 max_length, sampling_temp, keep_topk):
+        assert block_ngram_repeat == 0
+        super(GreedySearch, self).__init__(
+            pad, bos, eos, batch_size, 1, min_length, block_ngram_repeat,
+            exclusion_tokens, return_attention, max_length)
         self.sampling_temp = sampling_temp
         self.keep_topk = keep_topk
         self.topk_scores = None
-        self.memory_length = memory_length
         self.batch_size = batch_size
-        self.select_indices = torch.arange(self.batch_size,
-                                           dtype=torch.long, device=device)
+        self.select_indices = torch.arange(self.batch_size, dtype=torch.long)
         self.original_batch_idx = torch.arange(self.batch_size,
-                                               dtype=torch.long, device=device)
+                                               dtype=torch.long)
+
+    def _init_runtime(self, device, memory_lengths):
```

Comment:
Is it separate from `initialize` just because of how this class is tested? Would be better to merge it with `initialize` and simply create all attributes there.